soroban-sdk/src/set.rs

```diff
@@ -0,0 +1,276 @@
+use core::{cmp::Ordering, fmt::Debug};
+
+use super::{Env, IntoVal, Map, RawVal, TryFromVal};
+
+/// Create a [Set] with the given items.
+///
+/// The first argument in the list must be a reference to an [Env], then the
+/// items follow.
+///
+/// ### Examples
+///
+/// ```
+/// use soroban_sdk::{Env, set};
+///
+/// let env = Env::default();
+/// let set = set![&env, 0, 1, 2, 3, 3];
+/// assert_eq!(set.len(), 4);
+/// ```
+#[macro_export]
+macro_rules! set {
+    ($env:expr) => {
+        $crate::Set::new($env)
+    };
+    ($env:expr, $($x:expr),+ $(,)?) => {
+        $crate::Set::from_array($env, [$($x), +])
+    };
+}
+
+/// Set is a contiguous growable array type composed of unique items.
+///
+/// A Set can be seen as syntactic sugar on top of the soroban-sdk Map
+/// implementation, where all of the Set's items are keys in a (hash) Map.
+/// This forces all items to be unique, and the action of adding an
+/// already-existing item to the Map is an idempotent operation.
+///
+/// Set imposes a fixed ordering, so Set<1, 2, 3> == Set<2, 3, 1>.
+///
+/// The array is stored in the Host and available to the Guest through the
+/// functions defined on Set. Values stored in the Set are transmitted to the
+/// Host as [RawVal]s, and when retrieved from the Set are transmitted back and
+/// converted from [RawVal] back into their type.
+///
+/// The values in a Set are not guaranteed to be of type `T` and conversion will
+/// fail if they are not.
+///
+/// ### Examples
+///
+/// ```
+/// use soroban_sdk::{Env, set};
+///
+/// let env = Env::default();
+/// let mut set = set![&env, 1, 2, 3];
+/// set.insert(3);
+/// assert_eq!(set.len(), 3);
+/// ```
+pub struct Set<T>(Map<T, ()>);
+
+impl<T> Set<T>
+where
+    T: IntoVal<Env, RawVal> + TryFromVal<Env, RawVal>,
+{
+    pub(crate) fn env(&self) -> &Env {
+        self.0.env()
+    }
+
+    pub fn new(env: &Env) -> Set<T> {
+        let map = Map::<T, ()>::new(env);
```

Comment:
Fyi the type annotations shouldn't be required since they are derived from the destination.
```suggestion
        let map = Map::new(env);
```